### PR TITLE
MRG: Speed up EDF reading

### LINKED ISSUE #3852: BUG: Reading EDF files can be very slow
I am trying to read an EDF file, but the .read_data() (edit, i mean` load_data()`) function fails. Same thing happens with this [sample file](http://www.teuniz.net/edf_bdf_testfiles/eeg_recording.zip) from [here](http://www.teuniz.net/edf_bdf_testfiles/)

Looking a bit into the code I find that there is a call to `_read_segment_file` in `io\base.py` which raises an `NotImplementedError`. The Problem seems to be the call itself, it doesn't even get to the called functions inside. If I comment the call out, it works fine.
```
self._read_segment_file(data[:, this_sl], idx, fi,
                                    int(start_file), int(stop_file),
                                    cals, mult)
```

Strangely enough running the sample file from the EDF-API works without problems

I hope I provided all the info needed. Unfortunately I'm not very experienced in debugging functions or classes. Thanks!
```
Platform:      Windows-10-10.0.14393-SP0
Python:        3.5.2 (same in 2.7.x) (v3.5.2:4def2a2901a5, Jun 25 2016, 22:18:55) [MSC v.1900 64 bit (AMD64)]
Executable:    C:\Program Files\WinPython-3.5\python-3.5.2.amd64\pythonw.exe

mne:           0.14.dev0
numpy:         1.11.2 {lapack=C:/Program Files (x86)/IntelSWTools/compilers_and_libraries/windows/mkl/include, blas=C:/Program Files (x86)/IntelSWTools/compilers_and_libraries/windows/mkl/include}
scipy:         0.18.1
matplotlib:    1.5.3

sklearn:       0.18
nibabel:       Not found
nitime:        Not found
mayavi:        Not found
nose:          1.3.7
pandas:        0.19.1
pycuda:        Not found
skcuda:        Not found
```

### PR DESCRIPTION
**Pull request by Eric89GXL**
Closes #3852.

This decreases time to load [the sample file](http://www.teuniz.net/edf_bdf_testfiles/eeg_recording.zip) from over 8 seconds to under 2 seconds on my system.

@skjerns can you see if it's faster on your files?

### CODE REVIEW

**Comment by skjerns** [comment]
From 7.5 Minutes to 11 seconds! That's definitely something :)

Thanks, that was really quick.

**Comment by Eric89GXL** [comment]
Do the data still look okay? :)

(Our tests should really catch it if this broke something, but it's worth double-checking.)

**Comment by skjerns** [comment]
I've loaded it into MatLab using this script https://nl.mathworks.com/matlabcentral/fileexchange/47123-edfread2 and the values seem to be similar, but the MNE values are 1000000 too large.
so -0.297696020398689 in the Matlab-script becomes -297696.0203989165 with MNE. 

I faintly remember that I had this behaviour earlier already, but I'm not sure.

**Comment by Eric89GXL** [comment]
Yeah what I changed wouldn't have affected that. Can you open a separate issue for that? Sounds like we are off by a scale factor, likely uV vs V (but we are in uV whereas most readers give us V).

File: mne/io/edf/edf.py [hunk 1]
```diff
@@ -130,59 +130,68 @@ def _read_segment_file(self, data, idx, fi, start, stop, cals, mult):
...
-                                               npad=0)[r_sidx:r_eidx]
-                    this_data[ii, :n_buf_samp] = ch_data
-                data[:, d_sidx:d_eidx] = this_data[:, :n_buf_samp]
+                            # XXX resampling each chunk isn't great,
+                            # it forces edge artifacts to appear at
+                            # each buffer boundary :(
+                            # it can also be very slow...
+                            ch_data = resample(
+                                ch_data, buf_len, n_samps[ci], npad=0, axis=-1)
+                    assert ch_data.shape == (len(ch_data), buf_len)
```
Comment (agramfort) [inline mne/io/edf/edf.py]: use assert_true or assert_equal

**Comment by agramfort** [comment]
+1 for merge if @skjerns 

maybe @jaeilepp wants to have a look too

File: mne/io/edf/edf.py (hunk 1 above)
Comment (Eric89GXL) [inline mne/io/edf/edf.py]: This isn't a test, it's just a reminder / "active comment". We have this in a few other places already :)

**Comment by skjerns** [comment]
I digged a bit deeper into the issue and the slow down in comparison with the test file that I supplied seems to come from `_blk_read_lims`

The test file has `buf_len` of `20` (coming from `self._raw_extras[fi]['max_samp']` ) whereas my file has `max_samp` of `1`. Similarly `n_records` = `nsamples` in my file whereas `n_records ` = `nsamples` \ 20

Additionally, the [units] field is `1e-06` for the test file and `1` for my file, which explains the voltage issue (seems to be a problem with my file?).

**Comment by Eric89GXL** [comment]
Yeah the `buf_len = 1` would have been *much* slower than `buf_len = 20` in the old code, and approximately the same speed under the new code.

@jaeilepp do you have time to look?

**Comment by agramfort** [comment]
let's say it works :)

thx @Eric89GXL

**Pull request merged**